---
title: "Workflow management system"
author: "Dave Tang"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: false
editor_options:
  chunk_output_type: console
---

A workflow management system (WMS) is software that makes it easier to implement, execute, and manage workflows. The analysis of high-throughput sequencing (HTS) data benefits greatly from WMSs because the processing of raw HTS data requires many sequential steps. Typically in a bioinformatics workflow or pipeline, HTS data is analysed by quality control (QC) tools, preprocessed accordingly, aligned to a reference sequence, and summarised in a manner that is relevant to the biological question.

However, you may be wondering whether it is necessary to use a WMS if you can implement your bioinformatics pipeline using various shell (or similar scripting language) scripts in a relatively short amount of time. Is it worth investing your time learning yet another tool, especially if your interest is in the biology that the data holds? I think so because knowing at least one WMS is a core bioinformatic skill that everyone should possess.

If you're convinced that you should at least try to use a WMS you may be wondering how hard it is to learn to use one and how long it will take to implement your bioinformatics workflow with a WMS. I am more familiar with the [Workflow Description Language](https://github.com/openwdl/wdl) (WDL), which is a language that you can use to implement your workflow, so my comments are based on WDL. I don't think it is hard to learn and to use WDL; I wrote a short [blog post](https://davetang.org/muse/2020/01/09/learning-wdl/) on learning WDL, so you can be the judge.

With WDL, you set up each component of your pipeline as individual **tasks** and each task follows a defined structure; you can also set up specify resource usage per task, i.e. how much resource should be used for this task. You then create a workflow by calling your tasks; you can specify that the output of one task is used as input for the next task. A JSON file is used as a config file for your workflow and will specify parameters and the location of files, such as your input files. Finally, a separate tool called [Cromwell](https://cromwell.readthedocs.io/en/develop/) can be used to execute your pipeline and it is Cromwell that will handle all the logging, resource management, and pipeline execution. One of the benefits of WDL/Cromwell is that you can execute the same pipeline on different computational infrastructures making your pipeline very portable!

Another WMS is [Snakemake](https://snakemake.readthedocs.io/en/stable/index.html), which uses **rules** to set up each part of your bioinformatics pipeline, akin to **tasks** in WDL.

```
rule sort:
    input:
        "test.txt"
    output:
        "test.sorted.txt"
    shell:
        "sort -n {input} > {output}"
```

If we run the example above using Snakemake, the input file `test.txt` will get sorted numerically and the output is stored in `test.sorted.txt`. Typically, you would write a pipeline (a `Snakefile`) that takes input from a config file (e.g. `config.yaml`). If you wanted to run the pipeline for a new dataset, you will just need to create a new config file.

Other workflow management systems include [Bpipe](https://github.com/ssadedin/bpipe/) and [Nextflow](https://sciwiki.fredhutch.org/compdemos/nextflow/), which both are based on [Groovy](https://www.groovy-lang.org/). A survey conducted on Twitter has a list of other systems and [showed that Snakemake is the most popular](https://docs.google.com/forms/d/e/1FAIpQLScoj8Po4P3Qrh7rbJrq2R35c3PQsNCynEeEVUAdcGyly7TT_Q/viewanalytics). There is a [nice discussion on Reddit](https://www.reddit.com/r/bioinformatics/comments/a4fq4i/given_the_experience_of_others_writing/) on the strengths and weakness of different workflow management systems.

Personally I use WDL because that's what the Broad Institute uses and I wanted to use some of their pipelines. This is another advantage of learning a WMS because it is very likely that somebody has already implemented a similar pipeline. Check out these pipelines in using different WMSs:

* Snakemake - https://github.com/snakemake-workflows
* Nextflow - https://github.com/nextflow-io/awesome-nextflow
* WDL - https://github.com/biowdl

Even if your exact pipeline hasn't been previously implemented, you probably can still find a similar pipeline that you could modify for your needs. 

Lastly, WMSs make it much easier to reproducible your work. TBC.

## WDL

* Specification - https://github.com/openwdl/wdl/tree/main/versions
* Optional inputs - https://github.com/openwdl/wdl/blob/master/versions/1.0/SPEC.md#optional-inputs
* true and false - https://github.com/openwdl/wdl/blob/master/versions/1.0/SPEC.md#true-and-false

## Other links

* https://github.com/FredHutch/reproducible-workflows
